rust/core-lib/src/plugins.rs

```diff
@@ -0,0 +1,376 @@
+        // TODO: should check None
+    }
+
+    fn handle_request(&mut self, _ctx: RpcCtx<ChildStdin>, method: &str, params: &Value) ->
+        Result<Value, Value> {
+        let result = self.rpc_handler(method, params);
+        result.ok_or_else(|| Value::String("missing return value".to_string()))
+    }
+}
+
+impl<W: Write + Send + 'static> PluginRef<W> {
+    fn rpc_handler(&self, method: &str, params: &Value) -> Option<Value> {
+        let plugin_manager = {
+            self.0.lock().manager.upgrade()
+        };
+
+        if let Some(plugin_manager) = plugin_manager {
+            let cmd = serde_json::from_value::<PluginCommand>(params.to_owned())
+                .expect(&format!("failed to parse plugin rpc {}, params {:?}",
+                        method, params));
+            let result = plugin_manager.lock()
+                .handle_plugin_cmd(cmd, &self.0.lock().buffer_id);
+        result
+        } else {
+            None
+        }
+    }
+
+    /// Init message sent to the plugin.
+    pub fn init_buf(&self, buf_size: usize, rev: usize) {
+        let plugin = self.0.lock();
+        let params = json!({
+            "buf_size": buf_size,
+            "rev": rev,
+        });
+        plugin.peer.send_rpc_notification("init_buf", &params);
+    }
+
+    // TODO: send finer grain delta
+    // TODO: make this a synchronous request (but with a callback to not block),
+    // so editor can defer gc until request returns
+    /// Update message sent to the plugin.
+    pub fn update<F>(&self, update: &PluginUpdate, callback: F)
+            where F: FnOnce(Result<Value, Error>) + Send + 'static {
+        let params = serde_json::to_value(update).expect("PluginUpdate invalid");
+        self.0.lock().peer.send_rpc_request_async("update", &params, callback);
+    }
+}
+
+
+// example plugins. Eventually these should be loadd from disk.
+pub fn debug_plugins() -> Vec<PluginDescription> {
+    let mut path_base = env::current_exe().unwrap();
+    path_base.pop();
+    let make_path = |p: &str| -> PathBuf {
+        let mut pb = path_base.clone();
+        pb.push(p);
+        pb
+    };
+
+    vec![
+        PluginDescription::new("syntect", "0.0", make_path("xi-syntect-plugin")),
+        PluginDescription::new("braces", "0.0", make_path("bracket_example.py")),
+        PluginDescription::new("spellcheck", "0.0", make_path("spellcheck.py")),
+        PluginDescription::new("shouty", "0.0", make_path("shouty.py")),
+    ]
+}
+
+/// Describes attributes and capabilities of a plugin.
+///
+/// Note: - these will eventually be loaded from manifest files.
+#[derive(Debug, Clone)]
+pub struct PluginDescription {
+    pub name: String,
+    version: String,
+    //scope: PluginScope,
+    // more metadata ...
+    /// path to plugin executable
+    pub exec_path: PathBuf,
+}
+
+impl PluginDescription {
+    fn new<S, P>(name: S, version: S, exec_path: P) -> Self
+        where S: Into<String>, P: Into<PathBuf> {
```

Comment:
Generally the style I prefer is to put `{` on the next line by itself, same indent as `fn`